hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1677 @@
+"""
+An attempt at a simpler LST binner that makes more assumptions but runs faster.
+
+In particular, we assume that all baselines have the same time array and frequency array,
+and that each is present throughout the data array. This allows a vectorization.
+"""
+from __future__ import annotations
+
+import numpy as np
+from . import utils
+import warnings
+from pathlib import Path
+from .lstbin import config_lst_bin_files, sigma_clip
+from . import abscal
+import os
+from . import io
+import logging
+from hera_qm.metrics_io import read_a_priori_ant_flags
+from . import apply_cal
+from typing import Sequence
+import argparse
+from pyuvdata.uvdata.uvh5 import FastUVH5Meta
+from pyuvdata import utils as uvutils
+from .red_groups import RedundantGroups
+import h5py
+from functools import partial
+import yaml
+
+try:
+    profile
+except NameError:
+    def profile(fnc):
+        return fnc
+
+logger = logging.getLogger(__name__)
+
+@profile
+def simple_lst_bin(
+    data: np.ndarray,
+    data_lsts: np.ndarray,
+    baselines: list[tuple[int, int]],
+    lst_bin_edges: np.ndarray,
+    freq_array: np.ndarray,
+    flags: np.ndarray | None = None,
+    nsamples: np.ndarray | None = None,
+    rephase: bool = True,
+    antpos: np.ndarray | None = None,
+    lat: float = -30.72152,
+) -> tuple[np.ndarray, list[np.ndarray], list[np.ndarray], list[np.ndarray]]:
+    """
+    Split input data into a list of LST bins.
+
+    This function simply splits a data array with multiple time stamps into a list of
+    arrays, each containing a single LST bin. Each of the data arrays in each bin
+    are also rephased onto a common LST grid.
+
+    Parameters
+    ----------
+    data
+        The visibility data. Must be shape (ntimes, nbls, nfreqs, npols)
+    data_lsts
+        The LSTs corresponding to each of the time stamps in the data. Must have
+        length ``data.shape[0]``
```

Comment:
specify this is bin centers

Reply:
I added some documentation mentioning that it actually doesn't matter what the ``data_lsts`` represent. The key assumption is that the data is *treated* as if the integrations occur at a snapshot, i.e. a delta function in LST at the given ``data_lst``. Thus, the whole integration is put into one LST bin, regardless of whether in reality it might straddle between two bins. Whether the data LST is at the start, middle or end of the actual integration will incur slightly different errors in detail, but qualitatively similar.

Reply:
I added some documentation mentioning that it actually doesn't matter what the ``data_lsts`` represent. The key assumption is that the data is *treated* as if the integrations occur at a snapshot, i.e. a delta function in LST at the given ``data_lst``. Thus, the whole integration is put into one LST bin, regardless of whether in reality it might straddle between two bins. Whether the data LST is at the start, middle or end of the actual integration will incur slightly different errors in detail, but qualitatively similar.